syntax = "proto3";

import "google/protobuf/empty.proto";

enum BufferFormat {
  REAL = 0;
  COMPLEX = 1;
};

enum BufferDirection {
  ADC = 0;
  DAC = 1;
};

message BufferId {
  BufferDirection direction = 1;
  uint32 buffer_no = 2;
};

message SampledInfo {
  uint32 n_adc_buffers = 1;
  uint32 n_dac_buffers = 2;
};

message SampleBufferInfoQuery {
  BufferId buffer_id = 1;
};

message SampleBufferInfo {
  BufferFormat format = 1;
  uint32 nsamples = 2;
  float sample_rate = 3;
}

message SampleData {
  BufferId buffer_id = 1;
  uint32 nsamples = 2;
  bytes samples = 3; 
}

message ChannelEnable {
  BufferId buffer_id = 1;
  bool I = 2;
  bool Q = 3;
}

message GetSamplesRequest {
  BufferId buffer_id = 1;
}

message SetSamplesResponse {
  int32 result = 1;
}

message OneShot {
  BufferId buffer_id = 1;
  bool one_shot = 2;
}

message TriggerDelay {
  BufferId buffer_id = 1;
  uint32 delay = 2;
}

service PiRadioSampled {
  rpc GetSampledInfo(google.protobuf.Empty) returns (SampledInfo);
  rpc GetSampleBufferInfo(SampleBufferInfoQuery) returns (SampleBufferInfo);

  rpc SetSamples(SampleData) returns (SetSamplesResponse);
  rpc GetSamples(GetSamplesRequest) returns (SampleData);

  rpc SetOneShot(OneShot) returns (google.protobuf.Empty);
  rpc GetOneShot(BufferId) returns (OneShot);

  rpc SetTriggerDelay(TriggerDelay) returns (google.protobuf.Empty);
  rpc GetTriggerDelay(BufferId) returns (TriggerDelay);

  rpc GetChannelEnable(BufferId) returns (ChannelEnable);
  rpc SetChannelEnable(ChannelEnable) returns (google.protobuf.Empty);
  
  rpc Trigger(google.protobuf.Empty) returns (google.protobuf.Empty);
}
